pkg/gadgets/interface.go

```diff
@@ -108,33 +104,47 @@ type EventEnricherSetter interface {
 	SetEventEnricher(func(ev any) error)
 }
 
-type StartStopGadget interface {
-	Start() error
-	Stop()
-}
-
-// RunGadget is an alternative to StartStopGadget that expects the result to be emitted using the EventHandler
-// as well but doesn't need to be stopped (manually)
+// RunGadget is an interface for gadgets that are run in the background and emit
+// events as soon as they occur.
 type RunGadget interface {
-	Run() error
-}
+	// Run is expected to run the gadget and emits the events using the
+	// EventHandler. It is expected to be blocking and return only when the
+	// context is done, after which the gadget should clean up all resources.
+	// Notice that Run() will be called after operators are installed.
+	Run(GadgetContext) error
+}
+
+// RunWithResultGadget is an alternative to RunGadget that returns the result
+// of the gadget only at the end of its execution.
+type RunWithResultGadget interface {
+	// RunWithResult follows the same rules as Run() but instead of using an
+	// EventHandler to emit the events, it returns the result of the gadget as a
+	// byte array after the context is done.
+	RunWithResult(GadgetContext) ([]byte, error)
+}
+
+// InitRunClose is a gadget interface that must be initialized before the
+// operators are installed. This is useful when the gadget needs to be kept
+// up-to-date with the containers that need to be traced, as the operators will
+// start sending notifications of this as soon as it is installed.
+type InitRunClose interface {
```

Comment:
I think we should support this in a version with `RunWithResult` as well to keep it consistent. WDYT?

Having said that, I'd prefer to have an (optional) `InitClose` interface and then a combined mandatory `Run() ([]byte, error)` for both cases (with and without a result). I know not every gadget returns something here, but in that case you could just pass `nil` (and it's the same signature we use for `RunGadget()`). The advantage would be to have at least one interface that is bound to the `Gadget` interface (which otherwise goes back to being `any`).

Reply:
I try to see how it would look like in https://github.com/inspektor-gadget/inspektor-gadget/pull/1408/commits/e5a586a604794bd5dfe3d74cf67a45194b524df5.

Reply:
I was actually thinking the other way around: to remove `RunWithResult()` and only keep `Run()`, the result can be sent through the event handler. Could it work?

I feel like returning `[]byte` is very specific, so making it mandatory for all gadgets to return something seems weird, specially because they can't return any type but only `[]byte`.

Reply:
Hmm. The good thing about the `[]byte` is that there's a guarantee that there can only be one per gadget run. And it can be transformed to whatever output type is supported/requested afterwards (the transformers accept `[]byte` as input, which also makes it easy to transport using gRPC). Using the `EventHandler` for that feels weird, as there's no such guarantee.
The cleanest solution would probably be to have something similar to the `EventHandler`, but called `SetResult` with a `ResultSetter` interface. But I guess that would feel clunky, so this is IMHO the next best solution.

Reply:
I think we all have the same goal: Unify `Run` and `RunWithResult`. However, there are several options. So I suggest to unblock this PR and go ahead with both of them and unify them on a consecutive PR. WDYT?

Reply:
> which also makes it easy to transport using gRPC

If the returned types can be serialized to json / binary we're all good, aren't we? 

Some questions and worries I have about returning `[]byte`:

- How does the enrichment work in that case? 
- Can we use it for all snapshotters and profilers?
- What about users that use the tracers directly? Do they need to parse the returned `[]byte` to have a golang type to work on?

Reply:
> I suggest to unblock this PR and go ahead with both of them and unify them on a consecutive PR. WDYT?

I agree, let's unblock this one as soon as possible.